chia/wallet/puzzles/cat.clvm

```diff
@@ -0,0 +1,409 @@
+       (if (= (f condition) CREATE_COIN)
+         (c CREATE_COIN
+               (c (cat_puzzle_hash cat_mod_struct (f (r condition)))
+                    (r (r condition)))
+         )
+         (if (= (f condition) CREATE_COIN_ANNOUNCEMENT)
+           (list CREATE_COIN_ANNOUNCEMENT (concat 0xca (f (r condition))))
+           condition
+         )
+       )
+     )
+
+     ;; tweak all `CREATE_COIN` conditions, enforcing created coins to be cats
+     (defun morph_conditions (conditions cat_mod_struct)
+       (if conditions
+         (c
+           (morph_condition (f conditions) cat_mod_struct)
+           (morph_conditions (r conditions) cat_mod_struct)
+         )
+         ()
+       )
+     )
+
+     ;; given a coin triplet, return the id of the coin
+     (defun-inline coin_id_for_coin (coin cat_mod_struct)
+       (sha256 (f coin) (cat_puzzle_hash cat_mod_struct (f (r coin))) (f (r (r coin))))
+     )
+
+     ;; utility to fetch coin amount from coin
+     (defun-inline input_amount_for_coin (coin)
+       (f (r (r coin)))
+     )
+
+     ;; calculate the hash of an announcement
+     (defun-inline calculate_annoucement_id (this_coin_id this_subtotal next_coin_id cat_mod_struct)
+       ; NOTE: the next line containts a bug, as sha256tree1 ignores `this_subtotal`
+       (sha256 next_coin_id 0xcb this_coin_id this_subtotal)
+     )
+
+     ;; create the `ASSERT_COIN_ANNOUNCEMENT` condition that ensures the next coin's announcement is correct
+     (defun-inline create_assert_next_announcement_condition (this_coin_id this_subtotal next_coin_id cat_mod_struct)
+       (list ASSERT_COIN_ANNOUNCEMENT
+             (calculate_annoucement_id this_coin_id
+                                            this_subtotal
+                                            next_coin_id
+                                            cat_mod_struct
+             )
+       )
+     )
+
+     ;; here we commit to I_{k-1} and S_k
+     (defun-inline create_announcement_condition (prev_coin_id prev_subtotal)
+       (list CREATE_COIN_ANNOUNCEMENT
+             (concat 0xcb (concat prev_coin_id prev_subtotal))
+       )
+      )
+
+     ;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+     ;; this function takes a condition and returns an integer indicating
+     ;; the value of all output coins created with CREATE_COIN. If it's not
+     ;; a CREATE_COIN condition, it returns 0.
+
+     (defun-inline output_value_for_condition (condition)
+       (if (= (f condition) CREATE_COIN)
+         (f (r (r condition)))
+         0
+       )
+     )
+
+     ;; this function takes a list of conditions and returns an integer indicating
+     ;; the value of all output coins created with CREATE_COIN
+     (defun output_totals (conditions)
+       (if conditions
+         (+ (output_value_for_condition (f conditions)) (output_totals (r conditions)))
+         0
+       )
+     )
+
+     ;; add three conditions to the list of morphed conditions:
+     ;; ASSERT_MY_COIN_ID for `this_coin_info`
```

Comment:
This condition is no longer being added. The `this_coin_info` value is being passed in by the solution. How are we ensuring that it's correct? Can we prove that counterfeiting is impossible even if it's incorrect?

Reply:
It actually is being added just not here.  it probably should be here, or at the very least the comment should move.